tests/integration/git/test.js

```diff
@@ -0,0 +1,352 @@
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 Open Assessment Technologies SA;
+ */
+
+/**
+ * This module tests the src/git.js module against the repos provided in fixtures
+ *
+ * When running each test:
+ * - folders and files from `tests/integration/fixtures` will be copied into `tests/integration/work`
+ * - a git client will be used to set up 3 temporary git repos:
+ *   - remoteRepo (a bare repo)
+ *   - localRepo (1st working copy of remoteRepo)
+ *   - secondLocalRepo (2nd working copy of remoteRepo)
+ * - after the tests, the `work` folder will be cleared, destroying the temporary repos
+ *
+ * @author Martin Nicholson <martin@taotesting.com>
+ */
+
+const path = require('path');
+const fs = require('fs-extra');
+const replace = require('replace-in-file');
+const test = require('tape');
+
+const gitFactory = require('../../../src/git.js');
+const simpleGit = require('simple-git/promise');
+
+const localRepoFixturePath = path.resolve(__dirname, '../fixtures/localRepo');
+const remoteRepoFixturePath = path.resolve(__dirname, '../fixtures/remoteRepo');
+const secondLocalRepoFixturePath = path.resolve(__dirname, '../fixtures/secondLocalRepo');
+const workDir = path.resolve(__dirname, '../work');
+const localRepoPath = path.join(workDir, 'localRepo');
+const remoteRepoPath = path.join(workDir, 'remoteRepo');
+const secondLocalRepoPath = path.join(workDir, 'secondLocalRepo');
+
+/**
+ * Lifecycle functions
+ */
+const setUp = async function setUp() {
+
+    tearDown();
+    try {
+        // copy 3 folders from `fixtures` into work dir
+        fs.copySync(localRepoFixturePath, localRepoPath);
+        fs.copySync(remoteRepoFixturePath, remoteRepoPath); // empty
+        fs.copySync(secondLocalRepoFixturePath, secondLocalRepoPath); // empty
```

Comment:
Is it on purpose than you perform a sync copy ?

```js
await fs.copy(localRepoFixturePath, localRepoPath);
await fs.copyremoteRepoFixturePath, remoteRepoPath);
await fs.copy(secondLocalRepoFixturePath, secondLocalRepoPath);
```

Reply:
I know it *should* be equivalent to using `await fs.copy()`, but for some reason it all fails when I replace the synchronous `fs` calls with async ones.

The current way may be a mix of paradigms but it does run.